src/TropicalGeometry/curve.jl

```diff
@@ -42,26 +44,487 @@ end
+    n = Polymake.ncols(G)
+    newcoeff = Vector{Int}(coefficients(dtc))
+    S = Set{Int}(1:n)
+    W = setdiff(S,vertex)
+    @assert all(j -> newcoeff[j]>=0, W) "Divisor not effective outside the vertex number $vertex"
+    while !(isempty(W))
+	w0 = vertex
+	if all(w -> newcoeff[w] >= outdegree(tc,W,w),W)
+	    coeffdelta = zeros(Int, n)
+	    delta = DivisorOnTropicalCurve(tc,coeffdelta)
+	    for j in W 
+	        delta = chip_firing_move(delta,j)
+	    end
+	    newcoeff = newcoeff + coefficients(delta)
+	else 
+	    for w in W 
+	        if newcoeff[w] < outdegree(tc,W,w) 
+		    w0 = w
+		    break	    	   
+	        end
+	    end
+	    W = setdiff(W,w0)
+        end
+    end    
+    reduced = DivisorOnTropicalCurve(tc, newcoeff)
+    return reduced   
 end
 
+@doc Markdown.doc"""
+   is_linearly_equivalent(dtc1::DivisorOnTropicalCurve{M, EMB}, dtc2::DivisorOnTropicalCurve{M, EMB})     
+
+Given two effective divisors `dtc1` and `dtc2` on the same tropica curve checks whether they are linearly equivalent. 
+```jldoctest
+julia> IM = IncidenceMatrix([[1,2],[1,3],[1,4],[2,3],[2,4],[3,4]]);
+
+julia> tc = TropicalCurve{min}(IM)
+An abstract tropical curve
+
+julia> coeffs1 = [0, 1, 1, 1];
+
+julia> dtc1 = DivisorOnTropicalCurve(tc,coeffs1)
+DivisorOnTropicalCurve{min, false}(An abstract tropical curve, [0, 1, 1, 1])
+
+julia> coeffs2 = [3,0,0,0];
+
+julia> dtc2 = DivisorOnTropicalCurve(tc,coeffs2)
+DivisorOnTropicalCurve{min, false}(An abstract tropical curve, [3, 0, 0, 0])
 
+julia> is_linearly_equivalent(dtc1, dtc2)
+true
+```
+"""
+function is_linearly_equivalent(dtc1::DivisorOnTropicalCurve{M, EMB}, dtc2::DivisorOnTropicalCurve{M, EMB}) where {M, EMB}
+    @assert is_effective(dtc1) "The divisor $dtc1 is not effective"
+    @assert is_effective(dtc2) "The divisor $dtc2 is not effective"
+    @assert base_curve(dtc1) === base_curve(dtc2) "The input curve needs to be the same for both divisors"
+    v = 1
+    reduced1 = v_reduced(dtc1,v)
+    reduced2 = v_reduced(dtc2,v)
+    coeff1 = coefficients(reduced1)
+    coeff2 = coefficients(reduced2)
+    return coeff1 == coeff2 
+end
+
+export DivisorOnTropicalCurve,
+    n_nodes,
+    degree,
+    coefficients,
+    is_effective,
+    graph,
+    base_curve,
+    chip_firing_move,
+    v_reduced, 
+    is_linearly_equivalent,
+    StructureTropicalJacobian,
+    visualize
 
 ###
-# 3. Basic properties
+# 4. More properties
 # -------------------
 ###
+@doc Markdown.doc"""
+    ElementaryDivisors(M::MatrixElem{T}) where T <: Int
+
+Calculates the elementary divisors of the matrix M, together with their product. 
+This assumes that the matrix is square and of rank n-1. 
+"""
+function ElementaryDivisors(M::MatrixElem{T}) where T <: fmpz
+  @assert ncols(M)==nrows(M) "Not a square matrix."
+  C = snf(M)
+  c0 = 1
+  n = ncols(M)
+  Z = Vector{T}(undef,n)
+  for i in 1:n
+    Z[i] = C[i,i]
+    if i <= n-1
+      c0 = Z[i]*c0
+    end 
+  end 
+  return Z,c0
+end
+
+@doc Markdown.doc"""
```

Comment:
Lower case. Why not return an abelian group? Then one does not have to return this redundant order argument.

Reply:
Changed.